assets/src/edit-story/app/media/mediaProvider.js

```diff
@@ -91,6 +99,101 @@ function MediaProvider({ children }) {
     ]
   );
 
+  const resetMedia = useCallback(
+    ({ pagingNum: p = 1 } = {}) => {
+      fetchMediaStart({ pagingNum: p });
+      getMedia({ mediaType, searchTerm, pagingNum: p })
+        .then(({ data }) => {
+          setMedia({
+            media: data,
+          });
+        })
+        .catch(fetchMediaError);
+    },
+    [
+      setMedia,
+      fetchMediaError,
+      fetchMediaStart,
+      getMedia,
+      mediaType,
+      searchTerm,
+    ]
+  );
+
+  const uploadMediaFromLibrary = useCallback(
+    async (files) => {
+      try {
+        const localMedia = await Promise.all(
+          files.map(getResourceFromLocalFile).reverse()
+        );
+        const filesUploading = files.map((file) => uploadFile(file));
+        setMedia({
+          media: [...localMedia, ...media],
+        });
+        await Promise.all(filesUploading);
+
+        // To avoid race conditions updating media library, a new request is necessary
+        resetMedia({ pagingNum });
+      } catch (e) {
+        fetchMediaError(e);
+      }
+    },
+    [resetMedia, setMedia, fetchMediaError, uploadFile, media, pagingNum]
+  );
+
+  const uploadMediaFromWorkspace = useCallback(
+    async (files, { insertElement, updateElementById }) => {
+      try {
+        const filesOnCanvas = await Promise.all(
+          files
+            .map(async (file) => {
+              const resource = await getResourceFromLocalFile(file);
+              const element = insertElement(resource.type, {
+                resource: getResourceFromAttachment(resource),
+              });
+
+              return {
+                resource,
+                element,
+                file,
+              };
+            })
+            .reverse()
+        );
+
+        setMedia({
+          media: [...filesOnCanvas.map(({ resource }) => resource), ...media],
+        });
+
+        await Promise.all(
+          filesOnCanvas.map(async ({ element, file }) => {
+            const uploadedFile = await uploadFile(file);
+            const resource = getResourceFromUploadAPI(uploadedFile);
+
+            updateElementById({
+              elementId: element.elementId,
+              properties: {
+                resource: {
+                  ...resource,
+                  poster: resource.poster,
```

Comment:
What does this line do?

Reply:
It ensures that we always have a poster image. From what we had before, the API was not returning a poster image from recently uploaded images. I will take a look to check if it was changed after recent additions related to uploading videos.

Reply:
Posters have to be generated in client. See https://github.com/google/web-stories-wp/pull/553

Reply:
Wouldn't this be handled by the spread operator.